### ISSUE TITLE
out of bounds stack read in function HUF_readStats on malformed input

### ISSUE THREAD

**Issue by hannob** [Cyan4973/zstd#49]
This input file
https://crashes.fuzzing-project.org/zstd-oob-stack-HUF_readStats
causes an out of bounds stack read access in zstd. To see this one needs to compile zstd with address sanitizer (-fsanitize=address in CFLAGS).

Issue was found with the help of american fuzzy lop.

This is the output from address sanitizer:
```
==19506==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffef5269784 at pc 0x0000004faff7 bp 0x7ffef5269520 sp 0x7ffef5269518
READ of size 4 at 0x7ffef5269784 thread T0
    #0 0x4faff6 in HUF_readStats /f/zst/zstd/programs/../lib/huff0.c:612:9
    #1 0x4fa03d in HUF_readDTableX2 /f/zst/zstd/programs/../lib/huff0.c:644:13
    #2 0x501080 in HUF_decompress4X2 /f/zst/zstd/programs/../lib/huff0.c:859:17
    #3 0x5138d2 in HUF_decompress /f/zst/zstd/programs/../lib/huff0.c:1701:23
    #4 0x4e4b39 in ZSTD_decompressLiterals /f/zst/zstd/programs/../lib/zstd.c:1078:21
    #5 0x4e4b39 in ZSTD_decodeLiteralsBlock /f/zst/zstd/programs/../lib/zstd.c:1102
    #6 0x4e6bb5 in ZSTD_decompressBlock /f/zst/zstd/programs/../lib/zstd.c:1468:23
    #7 0x4e68b2 in ZSTD_decompressContinue /f/zst/zstd/programs/../lib/zstd.c:1622:21
    #8 0x52dcbf in FIO_decompressFrame /f/zst/zstd/programs/fileio.c:396:23
    #9 0x52e721 in FIO_decompressFilename /f/zst/zstd/programs/fileio.c:492:21
    #10 0x530fe3 in main /f/zst/zstd/programs/zstdcli.c:352:9
    #11 0x7f76943b6f9f in __libc_start_main /var/tmp/portage/sys-libs/glibc-2.20-r2/work/glibc-2.20/csu/libc-start.c:289
    #12 0x4377c6 in _start (/mnt/ram/zstd/zstd+0x4377c6)

Address 0x7ffef5269784 is located in stack of thread T0 at offset 420 in frame
    #0 0x4f9edf in HUF_readDTableX2 /f/zst/zstd/programs/../lib/huff0.c:630

  This frame has 4 object(s):
    [32, 288) 'huffWeight'
    [352, 420) 'rankVal' <== Memory access at offset 420 overflows this variable
    [464, 468) 'tableLog'
    [480, 484) 'nbSymbols'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /f/zst/zstd/programs/../lib/huff0.c:612 HUF_readStats
Shadow bytes around the buggy address:
  0x10005ea452a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005ea452b0: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
  0x10005ea452c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005ea452d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005ea452e0: f2 f2 f2 f2 f2 f2 f2 f2 00 00 00 00 00 00 00 00
=>0x10005ea452f0:[04]f2 f2 f2 f2 f2 04 f2 04 f3 f3 f3 00 00 00 00
  0x10005ea45300: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
  0x10005ea45310: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005ea45320: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005ea45330: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005ea45340: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==19506==ABORTING

**Comment by Cyan4973** [comment]
Thanks Hanno, this is exactly the kind of report which is needed right now. This issue requires a pretty impressive set of conditions to trigger (it must successfully pass a lot of prior check tests to get there).

I'm a bit surprised because I ran afl for almost 24h before release, without finding any issue.
I initially thought it was a matter of randomness. Digging deeper though, I note that when introducing your test case into the initial sef of afl, afl just runs normally, it doesn't find any crash.

Impressively enough, I also note that this bug pass _completely unnoticed by valgrind memory tester_, which is my tool of choice to grab such issues. Only clang + address sanitizer can catch it. So I'll reconsider its usage into my CI environment.

I suspect this might be because the faulty operation happens a few bytes beyond the limit of the intended table, but still within the limit of the local stack (there are some other tables after this one). 

Anyway, bottom line : only address sanitizer seems able to catch this issue.
I therefore tried to combine afl with address sanitizer (which comes with its own set of difficulties, due to virtual addressing), but  afl stops during initialization phase, while trying to decode zstd-oob-stack-HUF_readStats. So it doesn't even start fuzzing.

So the question is : how did you find this issue ?

**Comment by hannob** [comment]
Okay, a bit of background:
It doesn't surprise me that valgrind can't find these issues, because that is virtually impossible, it's just a valid access to another valid memory address. A stack overflow can't be found by any runtime tool, because the bare assembly code has no knowledge of what is valid and what not, a tool needs to change the memory layout at compile time to be able to detect these issues.

The "secret" of running afl+asan combined is that asan requires lots (terabytes) of virtual memory, so this only works if you disable the afl memory limit (-m none). Also see my tutorial:
https://fuzzing-project.org/tutorial3.html
There are more advanced tricks to make afl+asan work together using cgroups, but I never bothered to test them because disabling the memory limit works just fine for me (although this carries a small risk of making your system unstable if the fuzzer finds inputs that will make your tested app run wild and allocate lots of memory).

**Comment by Cyan4973** [comment]
Thanks for the link, this is a great read.

Indeed, using `-m none` is what I attempted to make `afl` work with asan, but another condition is that none of the initial seed files should make asan crash, otherwise fuzzing doesn't even start.

After that, fuzzing does proceed as expected, just quite a bit slower due to asan checks. I'm just curious about what happens when asan find another problem : does that make `afl` quit immediatly, or does it just add the file to the `crashes` directory ?

**Comment by Cyan4973** [comment]
The issue should be fixed into latest update of the "dev" branch

**Comment by hannob** [comment]
afl doesn't treat asan-crashes different from other crashes. So it will just collect crashes in the [outdir]/crashes directory. You'll usually get a couple of duplicates for the issues it finds.

**Comment by hannob** [comment]
also can confirm fix works, now testing with the new dev branch to see if it finds anything more.
